### PR TITLE
implement etcd backups

### PR DESCRIPTION
**Pull request by asalkeld**
Some notes:

- This vendors etcd and borrows some code from etcd-operator. I did this as etc operator has a very different version of azure client and was a nightmare to vendor.
- it creates 2 cron jobs a daily and an hourly and keeps 6 of each backups

TODO:
- [x] create quay.io/openshift-azure/backup
- [ ] e2e tests will be done separately
- [ ] we need to alert on failed backups and backups not running (again will be done separately)
- [x] This works great as a pod, but I am getting the following error as a cronjob (any help on getting around this is really appreciated)
Error creating: pods "etcd-backup-1541722800-" is forbidden: unable to validate against any security context constraint: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used]

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,13 +1,13 @@
 COMMIT=$(shell git rev-parse --short HEAD)$(shell [[ $$(git status --porcelain --ignored) = "" ]] && echo -clean || echo -dirty)
 
 # all is the default target to build everything
-all: clean build sync e2e-bin logbridge
+all: clean build sync e2e-bin logbridge backup
```
Comment (kwoodson) [inline Makefile]: suggestion: maybe update to etcd-backup as there might be other backup tools in the future

File: cmd/backup/backup.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	} else {
+		return nil
+	}
+	// this should sort oldest first
+	sort.SliceStable(blobs.Blobs, func(i, j int) bool {
+		return time.Time(blobs.Blobs[i].Properties.LastModified).Before(time.Time(blobs.Blobs[j].Properties.LastModified))
+	})
+	for _, blob := range blobs.Blobs {
+		logrus.Infof("pruning blob %v", blob.Name)
+		blob.Delete(nil)
```
Comment (jim-minter) [inline cmd/backup/backup.go]: error check?

File: cmd/backup/backup.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	if len(blobs.Blobs) > cMaxBackups {
+		toDelete = len(blobs.Blobs) - cMaxBackups
+	} else {
+		return nil
+	}
+	// this should sort oldest first
+	sort.SliceStable(blobs.Blobs, func(i, j int) bool {
+		return time.Time(blobs.Blobs[i].Properties.LastModified).Before(time.Time(blobs.Blobs[j].Properties.LastModified))
+	})
+	for _, blob := range blobs.Blobs {
```
Comment (jim-minter) [inline cmd/backup/backup.go]: for robustness I would be highly inclined to filter blobs to only the ones you want to delete, then iterate them - the toDelete/continue thing is kinda nasty.

File: cmd/backup/backup.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+const (
+	cMaxBackups = 6
+	copyRetries = 3
+	interval    = 5 * time.Second
+)
+
+var (
+	logLevel    = flag.String("loglevel", "Debug", "Valid values are Debug, Info, Warning, Error")
+	blobName    = flag.String("blobname", "", "Name of the blob")
```
Comment (jim-minter) [inline cmd/backup/backup.go]: can these be whittled down to `./backup save` and `./backup restore blobname >file` ? (don't care about the choice of words, just avoiding the options)

File: pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup-daily.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            image: '*** GENERATED ***'
+            imagePullPolicy: Always
+            command:
+            - /backup
+              -group daily
+            volumeMounts:
+            - name: azureconfig
+              mountPath: /_data/_out
+              readOnly: true
+          volumes:
```
Comment (jim-minter) [inline pkg/addons/data/CronJob.batch/openshift-etcd/etcd-backup-daily.yaml]: expect to see an etcd hostmount here as well.

**Comment by jim-minter** [comment]
@asalkeld:

> Error creating: pods "etcd-backup-1541722800-" is forbidden: unable to validate against any security context constraint: [spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used]

see pkg/addons/data/SecurityContextConstraints.security.openshift.io/privileged.yaml - you'll need to add the system:serviceaccount:openshift-etcd:etcd-operator service account for hostPath to work.

File: cmd/backup/backup.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	rev, etcdVersion, err := bm.SaveSnap(b.ctx, path)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot (%v)", err)
+	}
+	logrus.Infof("SaveSnap rev:%v etcdVersion:%v", rev, etcdVersion)
+	return nil
+}
+
+func (b *backupCmd) handleBackupPrune() error {
+	containerRef := b.abs.GetContainerReference("etcd")
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply cmd/backup/backup.go]: Should we move this to API package, and same with static pod names?

File: pkg/config/images.go [hunk 7]
```diff
@@ -161,6 +163,9 @@ func (g *simpleGenerator) selectContainerImages(cs *api.OpenShiftManagedCluster)
 	if g.pluginConfig.LogBridgeImage != "" {
 		cs.Config.Images.LogBridge = g.pluginConfig.LogBridgeImage
 	}
+	if g.pluginConfig.BackupImage != "" {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/config/images.go]: done

File: Makefile (hunk 1 above)
Comment (asalkeld) [reply Makefile]: done

File: pkg/api/plugin.go [hunk 8]
```diff
@@ -52,6 +52,7 @@ type DeployFn func(context.Context, map[string]interface{}) error
 type PluginConfig struct {
 	SyncImage       string
 	LogBridgeImage  string
+	BackupImage     string
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/api/plugin.go]: done

File: cmd/backup/backup.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup.go]: done

File: cmd/backup/backup.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		return err
+	}
+	if !containerExists {
+		return fmt.Errorf("container %v does not exist", "etcd")
+	}
+
+	params := azstorage.ListBlobsParameters{
+		Prefix:     "backup-" + *group,
+		MaxResults: cMaxBackups + 5,
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup.go]: ok, makes sense

File: cmd/backup/backup.go (hunk 2 above)
Comment (asalkeld) [reply cmd/backup/backup.go]: done

File: cmd/backup/tls.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 The etcd-operator Authors
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/tls.go]: ok, can do.

File: cmd/backup/retrieve.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+	blob := containerRef.GetBlobReference(key)
+	return blob.Get(&azstorage.GetBlobOptions{})
+}
+
+func (r *backupCmd) copy(srcBlob, destPath string) error {
+	logrus.Printf("copy blob %v to filesystem %v", srcBlob, destPath)
+	var rc io.ReadCloser
+	var err error
+	err = wait.PollImmediateInfinite(time.Second, func() (bool, error) {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/retrieve.go]: done

File: cmd/backup/retrieve.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	"os"
+	"strings"
+	"time"
+
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func parseBucketAndKey(path string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/retrieve.go]: done

File: cmd/backup/backup_manager.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+)
+
+// BackupManager backups an etcd cluster.
+type BackupManager struct {
+	kubecli kubernetes.Interface
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup_manager.go]: done

File: pkg/util/managedcluster/managedcluster.go [hunk 14]
```diff
@@ -52,6 +54,34 @@ func ClientsetFromV1Config(config *v1.Config) (*kubernetes.Clientset, error) {
 	return kubernetes.NewForConfig(restconfig)
 }
 
+func MustNewKubeClient() kubernetes.Interface {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/util/managedcluster/managedcluster.go]: done

File: cmd/backup/backup.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+	blobName    = flag.String("blobname", "", "Name of the blob")
+	group       = flag.String("group", "hourly", "The Blob will be called etcd/backup-<group>-<timestamp>")
+	destination = flag.String("destination", "", "Where to place the blob on the filesystem")
+	gitCommit   = "unknown"
+)
+
+type backupCmd struct {
+	abs     *azstorage.BlobStorageClient
+	ctx     context.Context
+	kubecli kubernetes.Interface
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup.go]: done

File: cmd/backup/backup.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	logLevel    = flag.String("loglevel", "Debug", "Valid values are Debug, Info, Warning, Error")
+	blobName    = flag.String("blobname", "", "Name of the blob")
+	group       = flag.String("group", "hourly", "The Blob will be called etcd/backup-<group>-<timestamp>")
+	destination = flag.String("destination", "", "Where to place the blob on the filesystem")
+	gitCommit   = "unknown"
+)
+
+type backupCmd struct {
+	abs     *azstorage.BlobStorageClient
+	ctx     context.Context
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup.go]: done

File: cmd/backup/backup_manager.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	"fmt"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+)
+
+// BackupManager backups an etcd cluster.
+type BackupManager struct {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup_manager.go]: done

File: cmd/backup/backup_manager.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+	defer rc.Close()
+
+	_, err = bm.bw.Write(ctx, s3Path, rc)
+	if err != nil {
+		return 0, "", fmt.Errorf("failed to write snapshot (%v)", err)
+	}
+	return rev, resp.Version, nil
+}
+
+// etcdClientWithMaxRevision gets the etcd endpoint with the maximum kv store revision
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/backup_manager.go]: the etcd's sync to each other so they can be at different revisions, it is finding the client with the newest change (so we backup the latest change):
```
		if resp.Header.Revision > maxRev {
			maxRev = resp.Header.Revision
			maxClient = etcdcli
		}

```

File: cmd/backup/abs_writer.go [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	blob := containerRef.GetBlobReference(key)
+	err = blob.CreateBlockBlob(&storage.PutBlobOptions{})
+	if err != nil {
+		return 0, err
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(r)
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/backup/abs_writer.go]: done

**Comment by asalkeld** [comment]
/hold
not tested, but most of the changes @jim-minter asked for.

File: cmd/etcdbackup/abs_writer.go [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+	// AzureBlobBlockChunkLimitInBytes 100MiB is the limit
+	AzureBlobBlockChunkLimitInBytes = 100 * 1024 * 1024
+)
+
+type blobWriter struct {
+	blob   *azstorage.Blob
+	blocks []azstorage.Block
+}
+
+var _ io.WriteCloser = &blobWriter{}
```
Comment (jim-minter) [inline cmd/etcdbackup/abs_writer.go]: for performance you also certainly need to implement io.ReaderFrom and read in at least 1MiB (I'd have thought) chunks.

File: cmd/etcdbackup/etcd_client.go [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+	"fmt"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/sirupsen/logrus"
+)
+
+// etcdClientWithMaxRevision gets the etcd endpoint with the maximum kv store revision
+// and returns the etcd client of that member.
+func (b *backupCmd) etcdClientWithMaxRevision(ctx context.Context) (*clientv3.Client, int64, error) {
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_client.go]: still needs simplification, as discussed on bluejeans

File: cmd/etcdbackup/etcd_backup.go [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/coreos/etcd/pkg/transport"
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/log"
+)
+
+const (
+	cMaxBackups = 6
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: we're not hungarian

File: cmd/etcdbackup/etcd_backup.go [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+)
+
+var (
+	logLevel    = flag.String("loglevel", "Debug", "Valid values are Debug, Info, Warning, Error")
+	blobName    = flag.String("blobname", "", "Name of the blob (without the container)")
+	group       = flag.String("group", "hourly", "The Blob will be called etcd/backup-<group>-<timestamp>")
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: surplus?

File: cmd/etcdbackup/etcd_backup.go [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+// saveSnap uses backup writer to save etcd snapshot to a specified path
+// and returns backup etcd server's kv store revision and its version.
+func (b *backupCmd) saveSnap(ctx context.Context, blobName string) (int64, string, error) {
+	etcdcli, rev, err := b.etcdClientWithMaxRevision(ctx)
+	if err != nil {
+		return 0, "", fmt.Errorf("create etcd client failed: %v", err)
+	}
+	defer etcdcli.Close()
+
+	resp, err := etcdcli.Status(ctx, etcdcli.Endpoints()[0])
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: needed?

File: cmd/etcdbackup/etcd_backup.go [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		return 0, "", fmt.Errorf("failed to receive snapshot (%v)", err)
+	}
+	defer rc.Close()
+
+	bw, err := newBlobWriter(b.abs, blobName)
+	if err != nil {
+		return 0, "", fmt.Errorf("failed to create blob (%v)", err)
+	}
+	defer bw.Close()
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: agreed not to use defer here as it's critical (presumably a mock unit test will be able to confirm this - open follow up if not implemented here)

File: cmd/etcdbackup/etcd_backup.go [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+
+type backupCmd struct {
+	abs           *azstorage.BlobStorageClient
+	endpoints     []string
+	etcdTLSConfig *tls.Config
+}
+
+// saveSnap uses backup writer to save etcd snapshot to a specified path
+// and returns backup etcd server's kv store revision and its version.
+func (b *backupCmd) saveSnap(ctx context.Context, blobName string) (int64, string, error) {
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: move into handleBackup()?

File: cmd/etcdbackup/etcd_backup.go [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+	} else {
+		return nil
+	}
+	// this should sort oldest first
+	sort.Slice(blobs.Blobs, func(i, j int) bool {
+		return time.Time(blobs.Blobs[i].Properties.LastModified).Before(time.Time(blobs.Blobs[j].Properties.LastModified))
+	})
+	for _, blob := range blobs.Blobs[:toDelete] {
+		logrus.Infof("pruning blob %v", blob.Name)
+		blob.Delete(nil)
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: error check

File: cmd/etcdbackup/etcd_backup.go [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+
+const (
+	cMaxBackups = 6
+)
+
+var (
+	logLevel    = flag.String("loglevel", "Debug", "Valid values are Debug, Info, Warning, Error")
+	blobName    = flag.String("blobname", "", "Name of the blob (without the container)")
+	group       = flag.String("group", "hourly", "The Blob will be called etcd/backup-<group>-<timestamp>")
+	destination = flag.String("destination", "", "Where to place the blob on the filesystem")
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: surplus?

File: cmd/etcdbackup/etcd_backup.go [hunk 29]
```diff
@@ -0,0 +1,206 @@
...
+
+	if flag.NArg() != 1 {
+		flag.Usage()
+	}
+
+	ctx := context.Background()
+	cli, err := getBlobStorageClient(ctx)
+	if err != nil {
+		logrus.Fatal(err)
+		os.Exit(1)
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: Fatal does this for you

File: cmd/etcdbackup/etcd_backup.go [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+	b := new(backupCmd)
+	b.abs = cli
+	b.endpoints = []string{
+		"https://master-000000:2379",
+		"https://master-000001:2379",
+		"https://master-000002:2379"}
+
+	switch flag.Arg(0) {
+	case "save":
+		path := fmt.Sprintf("backup-%s-%s", *group, time.Now().Format("2006-01-02T15-04-05"))
```
Comment (jim-minter) [inline cmd/etcdbackup/etcd_backup.go]: Specify UTC as well.

File: pkg/util/cloudproviderconf/cloudproviderconf.go [hunk 31]
```diff
@@ -0,0 +1,34 @@
+package cloudproviderconf
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/util/cloudproviderconf/cloudproviderconf.go]: done

File: pkg/util/cloudproviderconf/cloudproviderconf.go [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+)
+
+type CloudProviderConf struct {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/util/cloudproviderconf/cloudproviderconf.go]: done

File: pkg/api/types.go [hunk 33]
```diff
@@ -3,6 +3,8 @@ package api
 const (
 	// APIVersion is the version of this API
 	APIVersion = "internal"
+	// ETCDBackupContainer the ABS blob container that etcd backups are placed in.
+	ETCDBackupContainer = "etcd"
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/api/types.go]: done

File: pkg/api/config.go [hunk 34]
```diff
@@ -88,7 +88,8 @@ type ImageConfig struct {
 	WebConsole            string `json:"webConsole,omitempty"`
 	Console               string `json:"console,omitempty"`
 
-	LogBridge string `json:"logBridge,omitempty"`
+	LogBridge  string `json:"logBridge,omitempty"`
+	ETCDBackup string `json:"Backup,omitempty"`
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/api/config.go]: done

File: pkg/addons/data/ServiceAccount/openshift-etcd/etcd-operator.yaml [hunk 35]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: etcd-operator
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/addons/data/ServiceAccount/openshift-etcd/etcd-operator.yaml]: done

File: cmd/etcdbackup/abs_client.go [hunk 36]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/storage/mgmt/storage"
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/openshift-azure/pkg/util/cloudproviderconf"
+)
+
+func getBlobStorageClient(ctx context.Context) (*azstorage.BlobStorageClient, error) {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/abs_client.go]: done, note it is not in azureclients else i get recursive imports in the tests

File: cmd/etcdbackup/abs_writer.go [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/pborman/uuid"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+)
+
+const (
+	// AzureBlobBlockChunkLimitInBytes 100MiB is the limit
+	AzureBlobBlockChunkLimitInBytes = 100 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/abs_writer.go]: done

File: cmd/etcdbackup/abs_writer.go [hunk 38]
```diff
@@ -0,0 +1,60 @@
...
+func newBlobWriter(abs *azstorage.BlobStorageClient, blobName string) (*blobWriter, error) {
+	containerRef := abs.GetContainerReference(api.ETCDBackupContainer)
+	blob := containerRef.GetBlobReference(blobName)
+	err := blob.CreateBlockBlob(&azstorage.PutBlobOptions{})
+	if err != nil {
+		return nil, err
+	}
+	bw := blobWriter{
+		blob:   blob,
+		blocks: make([]azstorage.Block, 1)}
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/abs_writer.go]: yeah, it was a bug.

File: cmd/etcdbackup/abs_writer.go [hunk 39]
```diff
@@ -0,0 +1,60 @@
...
+	return &bw, nil
+}
+
+func (w *blobWriter) Write(data []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/abs_writer.go]: cool, thanks, that makes it simpler

File: cmd/etcdbackup/abs_writer.go [hunk 40]
```diff
@@ -0,0 +1,60 @@
...
+	dlen := len(data)
+	chunckCount := dlen/AzureBlobBlockChunkLimitInBytes + 1
+	for i := 0; i < chunckCount; i++ {
+		blockID := base64.StdEncoding.EncodeToString([]byte(uuid.New()))
+		w.blocks = append(w.blocks, azstorage.Block{ID: blockID, Status: azstorage.BlockStatusLatest})
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/abs_writer.go]: done

File: cmd/etcdbackup/etcd_client.go (hunk 21 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_client.go]: done

File: cmd/etcdbackup/etcd_backup.go (hunk 22 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: hehe

File: cmd/etcdbackup/etcd_backup.go (hunk 23 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: removed

File: cmd/etcdbackup/etcd_backup.go (hunk 24 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: removed

File: cmd/etcdbackup/etcd_backup.go (hunk 25 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: done

File: cmd/etcdbackup/etcd_backup.go (hunk 27 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: done

File: cmd/etcdbackup/etcd_backup.go (hunk 28 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: i have left it in for now..

File: cmd/etcdbackup/etcd_backup.go (hunk 29 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: ack

File: cmd/etcdbackup/etcd_backup.go (hunk 30 above)
Comment (asalkeld) [reply cmd/etcdbackup/etcd_backup.go]: done

File: cmd/etcdbackup/abs_writer.go (hunk 20 above)
Comment (asalkeld) [reply cmd/etcdbackup/abs_writer.go]: I ended up using an iobuf, i hope that's ok.

File: pkg/addons/translate.go [hunk 41]
```diff
@@ -301,6 +301,12 @@ var Translations = map[string][]struct {
 			Template:   "https://{{ .Derived.PublicHostname .ContainerService }}",
 		},
 	},
+	"CronJob.batch/openshift-etcd/etcd-backup-hourly.yaml": {
```
Comment (mjudeikis) [inline pkg/addons/translate.go]: first you dont need `.yaml` in the end.

Comment (asalkeld) [reply pkg/addons/translate.go]: doh! thanks

**Comment by asalkeld** [comment]
/cancel hold

ptal @jim-minter 
it's working on my cluster

File: pkg/util/accounthelper/accounthelper.go [hunk 42]
```diff
@@ -0,0 +1,153 @@
+package accounthelper
```
Comment (jim-minter) [inline pkg/util/accounthelper/accounthelper.go]: The refactoring into this file is much more than I was looking for and fundamentally I don't think it takes us in the right direction.  It doesn't feel like a good abstraction.  Can you largely revert this bit?  I think we're better keeping a bit of separation between plugin code and external containers - I definitely don't like moving bits of Initialize() in here.  I am happy for the etcdbackup container to do `c := bsc.GetContainerReference(EtcdBackupContainer); c.CreateIfNotExists(nil)` on its own if you want that.  All I'm really looking for is a lightweight way to share (not copy/paste) the code that was in `newAzureClients`.

File: cmd/etcdbackup/etcdbackup.go [hunk 43]
```diff
@@ -0,0 +1,118 @@
...
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/openshift-azure/pkg/etcdbackup"
+	"github.com/openshift/openshift-azure/pkg/log"
+	"github.com/openshift/openshift-azure/pkg/util/accounthelper"
+)
+
+const (
+	maxBackups = 6
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/etcdbackup.go]: done

File: cmd/etcdbackup/etcdbackup.go [hunk 44]
```diff
@@ -0,0 +1,118 @@
...
+		CertFile:      "/etc/origin/master/master.etcd-client.crt",
+		KeyFile:       "/etc/origin/master/master.etcd-client.key",
+		TrustedCAFile: "/etc/origin/master/master.etcd-ca.crt",
+	}
+	etcdTLSConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	cfg := clientv3.Config{
+		DialTimeout: 5 * time.Second,
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply cmd/etcdbackup/etcdbackup.go]: done

**Comment by jim-minter** [comment]
@asalkeld additionally please rebase, then we'll merge this :)

File: pkg/etcdbackup/etcdbackup.go [hunk 45]
```diff
@@ -0,0 +1,128 @@
...
+	defer rc.Close()
+
+	logrus.Infof("Creating New Blob")
+	bw, err := newBlobWriter(b.etcdContainer, bname)
+	if err != nil {
+		return errors.Wrap(err, "failed to create blob")
+	}
+
+	logrus.Infof("Copying blocks to blob")
+	bufferedBw := bufio.NewWriterSize(bw, azureBlobBlockChunkLimitInBytes/10)
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/etcdbackup/etcdbackup.go]: ha, today I learned about GRPC and its apparently 32KB default frame size ;)

[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/etcdbackup/etcdbackup.go]: bufio it is!

**Comment by jim-minter** [comment]
rebased and pushed, let's see if the unit test recurs

**Comment by kwoodson** [comment]
```
time="2018-11-14T19:15:28Z" level=debug msg="400 Bad Request: Failed to apply request: WaitForNodes: Get https://ci-op-jirxrlxd-8474e.eastus.cloudapp.azure.com/api/v1/nodes/master-000000: read tcp 172.16.186.133:56804->40.87.50.188:443: read: connection timed out" resourceGroup=ci-op-jirxrlxd-8474e
time="2018-11-14T19:15:28Z" level=fatal msg="containerservice.OpenShiftManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: error response cannot be parsed: \"400 Bad Request: Failed to apply request\\n\" error: json: cannot unmarshal number into Go value of type azure.RequestError" resourceGroup=ci-op-jirxrlxd-8474e
exit status 1
```

**Comment by kwoodson** [comment]
This deployment in the resource group stated that it was successful.
```
Your deployment is complete
```

**Comment by asalkeld** [comment]
this is going to need a rebase

**Comment by kargakis** [comment]
needs a rebase

/lgtm cancel

**Comment by kargakis** [comment]
All failures in the e2e test seem to be the same, the master VMs never come up.

https://openshift-gce-devel.appspot.com/pr/openshift_openshift-azure/713

/lgtm cancel

**Comment by mjudeikis** [comment]
I think there is some issue with e2e :/
/test e2e

**Comment by kargakis** [comment]
> I think there is some issue with e2e :/

Other PRs are testing fine

https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/730/pull-ci-openshift-openshift-azure-master-e2e-azure/769?log#log

**Comment by kwoodson** [comment]
Logging onto the cluster during the CI build results in the following:
```
[root@master-000000 ~]# oc get pods
NAME                        READY     STATUS             RESTARTS   AGE
controllers-master-000000   1/1       Running            0          10m
controllers-master-000001   1/1       Running            0          10m
controllers-master-000002   1/1       Running            0          10m
master-api-master-000000    1/1       Running            0          10m
master-api-master-000001    1/1       Running            0          10m
master-api-master-000002    1/1       Running            0          10m
master-etcd-master-000000   1/1       Running            0          10m
master-etcd-master-000001   1/1       Running            0          11m
master-etcd-master-000002   1/1       Running            0          11m
sync-master-000000          0/1       CrashLoopBackOff   6          11m
```

The sync pod logs:
```
[root@master-000000 ~]# docker logs 10fde834d1be
time="2018-11-16T16:06:28Z" level=info msg="sync pod starting, git commit d8953e2-clean"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xfc46e1]
goroutine 1 [running]:
main.(*sync).init(0xc4203d52c0, 0x131de80, 0xc420042030, 0x1, 0x1)
	/go/src/github.com/openshift/openshift-azure/cmd/sync/sync.go:46 +0x71
main.main()
	/go/src/github.com/openshift/openshift-azure/cmd/sync/sync.go:116 +0x137
```

**Pull request merged**